jwst/wavecorr/wavecorr.py

```diff
@@ -127,67 +129,82 @@ def apply_zero_point_correction(slit, reffile):
+    and add it to the WCS for each slit.
 
     Parameters
     ----------
     lam : ndarray
-        Wavelength array.
+        Wavelength array [in m].
+    dispersion : ndarray
+        The pixel dispersion [in m].
     freference : str
         ``wavecorr`` reference file name.
     source_xpos : float
         X position of the source as a fraction of the slit size.
     aperture_name : str
         Aperture name.
-    dispersion : ndarray
-        The pixel dispersion [in m].
-
+        
     Returns
     -------
-    lambda_corr : ndarray
-        Wavelength correction.
-    lam : ndarray
-        Interpolated wavelengths. Extrapolated values are reset to 0.
-        This is returned so that the DQ array can be updated with a flag
-        which indicates that no zero-point correction was done.
+    model : `~astropy.modeling.tabular.Tabular1D`
+        A model which takes wavelength inputs and returns zero-point
+        corrected wavelengths.
     """
+    # Open the zero point reference model
     with datamodels.WaveCorrModel(freference) as wavecorr:
         for ap in wavecorr.apertures:
             if ap.aperture_name == aperture_name:
                 log.info(f'Using wavelength zero-point correction for aperture {ap.aperture_name}')
                 offset_model = ap.zero_point_offset.copy()
-                # TODO: implement variance
-                # variance = ap.variance.copy()
-                # width = ap.width
                 break
         else:
             log.info(f'No wavelength zero-point correction found for slit {aperture_name}')
-
-    deltax = source_xpos
-    lam = lam.copy()
-    lam_no_nans = lam[~np.isnan(lam)]
+        
+    # Set lookup table to extrapolate at bounds to recover wavelengths
+    # beyond model bounds, particularly for the red and blue ends of
+    # prism observations.  fill_value = None sets the lookup tables
+    # to use the default extrapolation which is a linear extrapolation
+    # from scipy.interpolate.interpn
     offset_model.bounds_error = False
-    correction = offset_model(lam_no_nans * 10 ** -6, [deltax] * lam_no_nans.size)
-    lam[~np.isnan(lam)] = correction
-
-    # The correction for pixels outside the slit and wavelengths
-    # outside the wave_range is 0.
-    lam[np.isnan(lam)] = 0.
-    lambda_cor = dispersion * lam
-    return lambda_cor, lam
+    offset_model.fill_value = None
+        
+    # Average the wavelength and dispersion across 2D extracted slit and remove nans
+    # So that we have a 1D wavelength array for building a 1D lookup table wcs transform
+    lam_mean = np.nanmean(lam, axis=0)
+    disp_mean = np.nanmean(dispersion, axis=0)
+    nan_lams = np.isnan(lam_mean) | np.isnan(disp_mean)
+    lam_mean = lam_mean[~nan_lams]
+    disp_mean = disp_mean[~nan_lams]
+    
+    # Calculate the corrected wavelengths
+    pixel_corrections = offset_model(lam_mean, source_xpos)
+    lam_corrected = lam_mean + (pixel_corrections * disp_mean)
```

Comment:
The old method was subtracting the correction, and the difference between subtracting and adding the correction should the be source of these large differences.  When reviewing the zero-point corrections the NRS team noted that the correction should be added rather than subtracted and we confirmed this looking at observed data.  So the plus sign is an intentional update.

Reply:
Just to confirm what Chris said... The NIRSpec cal team looked at this and is convinced that the minus sign was an error, and we should use the plus sign going forward.